src/core/lwt.ml

```diff
@@ -22,1307 +22,2746 @@
  * 02111-1307, USA.
  *)
 
-(* +-----------------------------------------------------------------+
-   | Types                                                           |
-   +-----------------------------------------------------------------+ *)
 
-exception Canceled
 
-module Int_map = Map.Make(struct type t = int let compare = compare end)
+(* Reading guide
 
-type data = (unit -> unit) Int_map.t
-  (* Type of data carried by threads *)
+   Welcome to the implementation of the Lwt core! This is a big file, but we
+   hope that reading it (parts at a time!) will not be scary :) Here is why:
+
+
+   * Sectioning
+
+   The code is broken up into sections, each one of which is an internal module.
+   Most of the modules have a signature, which serves as a neat table of
+   contents.
+
+   It is recommended that you read this file with code folding enabled. If you
+   fold all the modules, you can visualize the logical structure of Lwt quite
+   easily. You can then expand modules as needed, depending on what part of the
+   implementation you are interested in. Without code folding, you face an
+   intimidating wall of code :( You can still visually parse the file, however,
+   because there are plenty of blank lines to help section things off.
+
+   The signatures are unusual: big comments are absent. They are moved into the
+   modules, so that they are hidden by code folding when you (the reader!) are
+   not interested in those modules.
+
+
+   * Documentation
+
+   The documentation begins with an Overview of major concepts and components.
+   This overview puts everything into context. You don't have to read the whole
+   thing. The overview begins with basic concepts, moves on to advanced ones,
+   and then gets into the truly esoteric. You can read about each concept on an
+   as-needed basis. However, once you have read the whole overview, you will be
+   aware of *everything* that is needed to understand, and work with, the core
+   of Lwt.
```

Comment:
IIUC the change in terminology (e.g. 'resolver' instead of 'wakener') comes from moving to 'promises' instead of 'threads'. Might be useful to add some table at the end on what the equivalent of the old `lwt.mli` terminology is in the new one for those who got used to the old terminology already?